OpenRA.Game/Download.cs

```diff
@@ -11,86 +11,53 @@
 
 using System;
 using System.ComponentModel;
+using System.IO;
 using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace OpenRA
 {
-	public class Download
-	{
-		readonly object syncObject = new object();
-		WebClient wc;
-
-		public static string FormatErrorMessage(Exception e)
-		{
-			var ex = e as WebException;
-			if (ex == null)
-				return e.Message;
-
-			switch (ex.Status)
-			{
-				case WebExceptionStatus.RequestCanceled:
-					return "Cancelled";
-				case WebExceptionStatus.NameResolutionFailure:
-					return "DNS lookup failed";
-				case WebExceptionStatus.Timeout:
-					return "Connection timeout";
-				case WebExceptionStatus.ConnectFailure:
-					return "Cannot connect to remote server";
-				case WebExceptionStatus.ProtocolError:
-					return "File not found on remote server";
-				default:
-					return ex.Message;
-			}
-		}
+	public delegate void OnProgress(long total, long totalRead, int progress);
 
-		void EnableTLS12OnWindows()
-		{
-			// Enable TLS 1.2 on Windows: .NET 4.7 on Windows 10 only supports obsolete protocols by default
-			// SecurityProtocolType.Tls12 is not defined in the .NET 4.5 reference dlls used by mono,
-			// so we must use the enum's constant value directly
-			if (Platform.CurrentPlatform == PlatformType.Windows)
-				ServicePointManager.SecurityProtocol |= (SecurityProtocolType)3072;
-		}
-
-		public Download(string url, string path, Action<DownloadProgressChangedEventArgs> onProgress, Action<AsyncCompletedEventArgs> onComplete)
-		{
-			EnableTLS12OnWindows();
-
-			lock (syncObject)
-			{
-				wc = new WebClient { Proxy = null };
-				wc.DownloadProgressChanged += (_, a) => onProgress(a);
-				wc.DownloadFileCompleted += (_, a) => { DisposeWebClient(); onComplete(a); };
-				wc.DownloadFileAsync(new Uri(url), path);
-			}
-		}
-
-		public Download(string url, Action<DownloadProgressChangedEventArgs> onProgress, Action<DownloadDataCompletedEventArgs> onComplete)
+	public static class HttpExtension
+	{
+		public static async Task ReadAsStreamWithProgress(this HttpResponseMessage response, Stream stream, OnProgress progress, CancellationToken token)
 		{
-			EnableTLS12OnWindows();
-
-			lock (syncObject)
-			{
-				wc = new WebClient { Proxy = null };
-				wc.DownloadProgressChanged += (_, a) => onProgress(a);
-				wc.DownloadDataCompleted += (_, a) => { DisposeWebClient(); onComplete(a); };
-				wc.DownloadDataAsync(new Uri(url));
-			}
-		}
+			var total = response.Content.Headers.ContentLength.HasValue
+				? response.Content.Headers.ContentLength.Value
+				: -1L;
+			var canReportProgress = total != -1;
 
-		void DisposeWebClient()
-		{
-			lock (syncObject)
+			using (var contentStream = await response.Content.ReadAsStreamAsync())
```

Comment:
`ReadAsStreamAsync()` has an overload with cancellation support:
```suggestion
			using (var contentStream = await response.Content.ReadAsStreamAsync(token))
```

Reply:
Wont work in mono sadly, the one with the cancellation token was added in core :(